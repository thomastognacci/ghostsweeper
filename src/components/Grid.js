import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Cell from "./Cell";

const StyledGrid = styled.div`
  margin: 15%;
  display: grid;
  grid-template-columns: ${(props) => `repeat(${props.cols}, 50px)`};
  grid-auto-rows: 50px;
  justify-content: center;
  align-items: center;

  > div:nth-child(odd) {
    background: #0080003b;
  }
  > div:nth-child(even) {
    background: #ff00003d;
  }
`;
class Grid extends React.PureComponent {
  state = {
    grid: [],
    cols: 5,
    rows: 5,
  };

  make2DArray = (cols, rows) => {
    const arr = new Array(cols);
    for (var i = 0; i < arr.length; i++) {
      arr[i] = new Array(rows);
    }
    return arr;
  };

  generateGrid = () => {
    const {cols, rows} = this.state;

    const grid = this.make2DArray(cols, rows);

    for (let i = 0; i < cols; i++) {
      for (let j = 0; j < rows; j++) {
        grid[i][j] = <Cell key={j + i * cols} x={j} y={i} />;
      }
    }

    this.setState({grid});
  };

  renderGrid = () => {
    const {grid} = this.state;

    return grid;
  };
  componentDidMount() {
    this.generateGrid();
  }
  render() {
    const {grid, cols, rows} = this.state;
    return (
      <StyledGrid cols={cols} rows={rows}>
        {grid && this.renderGrid()}
      </StyledGrid>
    );
  }
}

Grid.propTypes = {};

export default Grid;
