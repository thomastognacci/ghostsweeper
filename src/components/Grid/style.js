import styled from "styled-components";

export const StyledGrid = styled.div`
  display: grid;
  grid-auto-flow: column;
  grid-template-rows: ${(props) => `repeat(${props.rows}, ${500 / props.rows}px)`};
  grid-auto-columns: ${(props) => `${500 / props.rows}px`};
  justify-content: center;
  align-items: center;

  position: relative;
  width: 500px;
  margin: 10vh auto;
  border-radius: 10px;
  overflow: hidden;
  box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.25);

  &:before {
    content: "";
    position: absolute;
    pointer-events: none;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background-image: url("/img/bg.png");
    opacity: 0.35;
  }
`;
