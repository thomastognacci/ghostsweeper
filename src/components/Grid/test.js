import React from "react";
import {shallow} from "enzyme";
import renderer from "react-test-renderer";

import Grid from ".";

describe("<Grid />", () => {
  it("renders without crashing", () => {
    shallow(<Grid rows={5} cols={5} ghost={5} />);
  });

  it("renders a correct 5*5 grid", () => {
    const rendered = renderer.create(<Grid rows={5} cols={5} ghost={5} />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
