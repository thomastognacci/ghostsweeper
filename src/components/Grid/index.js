import React from "react";
import PropTypes from "prop-types";

import {StyledGrid} from "./style";
import Cell from "../Cell/";

class Grid extends React.PureComponent {
  state = {
    grid: [],
    mineAmount: this.props.mineAmount,
    cols: this.props.cols,
    rows: this.props.rows,
  };

  make2DArray = (cols, rows) => {
    const arr = new Array(cols);
    for (var i = 0; i < arr.length; i++) {
      arr[i] = new Array(rows);
    }
    return arr;
  };

  generateGrid = () => {
    const {cols, rows, mineAmount} = this.state;

    const grid = this.make2DArray(cols, rows);
    let options = [];
    for (let i = 0; i < cols; i++) {
      for (let j = 0; j < rows; j++) {
        let x = i;
        let y = j;
        options.push([x, y]);

        grid[x][y] = {
          key: j + x * cols,
          revealed: false,
          mine: false,
          mineCount: 0,
          x,
          y,
        };
      }
    }

    const gridWithMine = this.placeMines(mineAmount, grid, options);
    const gridWithCounter = this.countMines(gridWithMine);

    this.setState({grid: gridWithCounter});
  };

  placeMines = (amount, grid, options) => {
    let gridWithMine = grid;

    for (let i = 0; i < amount; i++) {
      let index = Math.floor(Math.random() * options.length);
      let pick = options[index];
      let x = pick[0];
      let y = pick[1];

      gridWithMine[x][y].mine = true;

      options.splice(index, 1);
    }

    return gridWithMine;
  };

  countMines = (grid) => {
    const {cols, rows} = this.props;

    for (let i = 0; i < cols; i++) {
      for (let j = 0; j < rows; j++) {
        let x = i;
        let y = j;

        if (grid[x][y].mine) {
          grid[x][y].mineCount = -1;
          continue;
        }
        let counter = 0;
        for (let xOff = -1; xOff <= 1; xOff++) {
          for (let yOff = -1; yOff <= 1; yOff++) {
            if (x + xOff > -1 && x + xOff < cols && y + yOff > -1 && y + yOff < rows) {
              let neighbour = grid[x + xOff][y + yOff];
              if (neighbour.mine) {
                counter++;
              }
            }
          }
        }
        grid[x][y].mineCount = counter;
      }
    }

    return grid;
  };

  handleCellClick = (x, y) => {
    const {grid} = this.state;

    if (grid[x][y].revealed) return;

    let newGrid;
    if (grid[x][y].mineCount === 0) {
      newGrid = this.revealNeighbor(grid, x, y);
    } else {
      newGrid = grid;
      newGrid[x][y].revealed = true;
    }
    this.setState({grid: [...newGrid]});
  };

  revealNeighbor = (grid, x, y) => {
    const {cols, rows} = this.props;

    for (let xOff = -1; xOff <= 1; xOff++) {
      for (let yOff = -1; yOff <= 1; yOff++) {
        if (x + xOff > -1 && x + xOff < cols && y + yOff > -1 && y + yOff < rows) {
          let neighbour = grid[x + xOff][y + yOff];
          if (!neighbour.revealed && !neighbour.mine) {
            neighbour.revealed = true;
            if (neighbour.mineCount === 0) {
              this.revealNeighbor(grid, x + xOff, y + yOff);
            }
          }
        }
      }
    }

    return grid;
  };

  renderGrid = () => {
    const {grid} = this.state;

    if (grid.length === 0) return;

    return grid.map((cols) => {
      return cols.map((cell) => {
        return (
          <Cell
            key={cell.key}
            x={cell.x}
            y={cell.y}
            mine={cell.mine}
            mineCount={cell.mineCount}
            revealed={cell.revealed}
            handleCellClick={this.handleCellClick}
          />
        );
      });
    });
  };
  componentDidMount() {
    this.generateGrid();
  }
  componentDidUpdate() {
    // console.log("up");
    // this.renderGrid();
  }
  render() {
    const {grid, cols, rows} = this.state;
    return (
      <StyledGrid onClick={this.handleClick} cols={cols} rows={rows}>
        {grid && this.renderGrid()}
      </StyledGrid>
    );
  }
}

Grid.propTypes = {
  cols: PropTypes.number.isRequired,
  rows: PropTypes.number.isRequired,
};

export default Grid;
