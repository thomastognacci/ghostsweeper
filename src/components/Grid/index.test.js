import React from "react";
import {shallow} from "enzyme";
import renderer from "react-test-renderer";

import Grid from "./";

// This prevent the randomness of the grid output
const mockMath = Object.create(global.Math);
mockMath.random = () => 0.5;
global.Math = mockMath;

describe("<Grid />", () => {
  it("renders without crashing", () => {
    shallow(<Grid />);
  });

  it("renders a correct 5*5 grid", () => {
    const rendered = renderer.create(<Grid rows={5} cols={5} />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
