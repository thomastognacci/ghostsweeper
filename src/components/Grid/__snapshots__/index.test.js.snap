// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> renders a correct 5*5 grid 1`] = `
<div
  className="sc-bdVaJa duDuyq"
  cols={5}
  rows={5}
>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
  <div
    className="land sc-bwzfXH kOGRNJ"
    onClick={[Function]}
  >
    
  </div>
</div>
`;
